src/django_components/slots.py

```diff
@@ -152,14 +214,120 @@ def __repr__(self) -> str:
 
     def render(self, context: Context) -> SafeString:
         trace_msg("RENDR", "SLOT", self.trace_id, self.node_id)
-        slots: Dict[SlotId, "SlotFill"] = context[_FILLED_SLOTS_CONTENT_CONTEXT_KEY]
-        # NOTE: Slot entry MUST be present. If it's missing, there was an issue upstream.
-        slot_fill = slots[self.node_id]
 
-        name, kwargs = self.resolve_kwargs(context)
+        if _COMPONENT_SLOT_CTX_CONTEXT_KEY not in context or not context[_COMPONENT_SLOT_CTX_CONTEXT_KEY]:
+            raise TemplateSyntaxError(
+                "Encountered a SlotNode outside of a ComponentNode context. "
+                "Make sure that all {% slot %} tags are nested within {% component %} tags.\n"
+                f"SlotNode: {self.__repr__()}"
+            )
+
+        component_ctx: ComponentSlotContext = context[_COMPONENT_SLOT_CTX_CONTEXT_KEY]
+        slot_name, kwargs = self.resolve_kwargs(context, component_ctx.component_name)
+
+        # Check for errors
+        if self.is_default and not component_ctx.is_dynamic_component:
```

Comment:
The `default` and `required` flags of the `{% slot %}` tag are now handled at render time